### PR TITLE
Change API to be cross-platform.

### PR DESCRIPTION
**Pull request by lrhn**
Changes the API to one that can efficiently be used to distinguish different target platforms, and which works both natively and on the web.

The `Platform` object no longer carry the native-only members, instead it *contains* either a `NativePlatform` object or a `BrowserPlatform` object, each providing operations suitable for the platform that is available.

Existing API still exists as deprecated, and migration from existing API is handled through `fix_data.yaml` as far as possible. (Currently not able to do complete migration, will work on enabling missing functionality.)

The API is intended to be scalable. If for example we need to handle `Wasm` as a third platform, adding a third `WasmPlatform` property to the `Platform` object, which only has a value if running on Wasm, should be a non-breaking extension.

Still suppports mocking for testing using a new `FakePlatform`, now carrying a `FakeNativePlatform` or `FakeBrowserPlatform`. The test classes need to be explicitly imported, to ensure that, fx, a web compiler can easily see that there is no concrete class implementing `NativePlatform` in the program when doing a production build, so `null`-checks against `Platform.current.nativePlatform` can be compile-time decided to enable better tree-shaking.

(Still missing the final migration, but please do review the intended API.)
@mit-mit

### CODE REVIEW

File: lib/platform.dart [hunk 1]
```diff
@@ -1,8 +1,41 @@
...
+
+/// Properties of the current platform.
+///
+/// The [Platform.current] object represents the current runtime platform
+/// by having one of [Platform.nativePlatform] or [Platform.browserPlatform]
+/// being non-`null`, depending on the runtime system the current program is
+/// running on.
+///
+/// That value can the provide more information about the current native,
+/// or browser platform.
```
Comment (goderbauer) [inline lib/platform.dart]: nit: something is broken in this sentence

File: lib/platform.dart [hunk 2]
```diff
@@ -1,8 +1,41 @@
...
+///
+/// [!NOTE]
+/// This library currently provides a deprecated legacy [LocalPlatform]
```
Comment (goderbauer) [inline lib/platform.dart]:
```suggestion
/// This library currently provides deprecated legacy [LocalPlatform]
```

File: lib/platform.dart [hunk 3]
```diff
@@ -1,8 +1,41 @@
...
+/// and use `FakeNativePlatform` instead.
+/// Be aware that `package:platform/testing.dart` exposes a
+/// *different* class named `FakePlatform`, so using the two libraries
+/// together requires hiding `FakePlatform` from `testing.dart`
+/// until all legacy [FakePlatform] uses have been removed.
+/// The legacy [FakePlatform] is also available as [LegacyFakePlatform].
+library;
+
+import 'src/legacy_implementation/legacy_classes.dart'; // For DartDoc.
+import 'src/platforms.dart'; // For DartDoc.
```
Comment (goderbauer) [inline lib/platform.dart]: could use the new `/// @docImport`s instead.

File: example/platform_using_switch.dart [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
```
Comment (goderbauer) [inline example/platform_using_switch.dart:1]: Could you also include an example test that showcases how to do mocking?

File: lib/src/platform_specific/platform_browser_interface.dart [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+/// Only available while running in a browser.
+abstract final class BrowserPlatform {
+  const BrowserPlatform._();
+
+  /// The current Browser platform, if any.
+  ///
+  /// Same as [Platform.current.browserPlatform].
+  static BrowserPlatform? get current => Platform.current.browserPlatform;
+
+  /// The browser's version, as reported by (something).
```
Comment (goderbauer) [inline lib/src/platform_specific/platform_browser_interface.dart]: nit: something?

File: lib/testing.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+/// Using the `run` methods of those fake-classes will run a function in
+/// a context where [Platform.current] exposes the fake configuration.
+///
+/// These fake classes should *only* be used for testing.
+/// Even including them in a production program may cause more code to
+/// be retained than necessary.
+library;
+
+import 'src/platforms.dart'; // For DartDoc.
+import 'src/testing/fake_platforms.dart'; // For DartDoc.
```
Comment (goderbauer) [inline lib/testing.dart]: `@docImport`s?

File: lib/testing.dart [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Platform information, and mocking of that information for tests.
+///
+/// This library provides the [Platform] interface of the `platform.dart`
+/// library, and the [BrowserPlatform] and [NativePlatform]
+/// types that it may or may not expose an instance of.
```
Comment (goderbauer) [inline lib/testing.dart:9]: Can you specify when it does and when it doesn't?

File: lib/testing.dart [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+///
+/// These classes can be used to specify a platform configuration that is not
+/// the actual current platform.
+/// Using the `run` methods of those fake-classes will run a function in
+/// a context where [Platform.current] exposes the fake configuration.
+///
+/// These fake classes should *only* be used for testing.
+/// Even including them in a production program may cause more code to
+/// be retained than necessary.
+library;
```
Comment (goderbauer) [inline lib/testing.dart:26]: Can we guard against this by making the entire library `@visibleForTesting`? If that doesn't work yet, might be worthwhile to file a request to extend `visibleForTesting` functionality to that.

**Comment by goderbauer** [comment]
What are the expectation around this change for the Flutter framework? Currently, the only place we access `Platform` from `dart:io` is in https://github.com/flutter/flutter/blob/fbd6890fb7bc29097ac7a40034ed60a6513f5194/packages/flutter/lib/src/foundation/_platform_io.dart. Can we continue to do that so we don't have to lock our users into a specific version of `pkg:platform`?

**Comment by lrhn** [comment]
I don't think the Flutter framework has to change anything. There is no problem importing `dart:io` if your code is intended to be native-only, or if the importing library is guarded by an `if (dart.library.io)`.
If your code is already cross-platform compatible, it's fine.

The target audience are people who currently import `dart:io` to do fx `Platform.isAndroid`, but who wants their code to also compile for the web.
They can change their code to instead import `package:platform/platform.dart` and do `Platform.currentPlatform.isAndroid`, and it will still *compile* for the web. It'll just return `false` for any check except `.isBrowser`.

The package is built and annotated with `@pragma`s to make it enable platform-specific compilations to recognize an `if (Platform.currentPlatform.isAndroid)` as, fx, always false when compiling for the web, or always true when actually compiling for Android, and discard the entire dead code branch. That can allow tree-shaking more code that was only called from platform specific branches.
That seems to work for both VM compiling with a `--target-os` flag and dart2js compiling to web.

File: lib/testing.dart (hunk 8 above)
Comment (lrhn) [reply lib/testing.dart:26]: I've added `@visibleForTesting` to each public declaration.
I can add it for the library too. (I don't know if it works, but it's worth a shot.)

File: lib/testing.dart (hunk 7 above)
Comment (lrhn) [reply lib/testing.dart:9]: Will do.

File: lib/testing.dart (hunk 6 above)
Comment (lrhn) [reply lib/testing.dart]: Hadn't landed when I wrote this, but will absolutely use it if it has.

File: lib/src/platform_specific/platform_browser_interface.dart (hunk 5 above)
Comment (lrhn) [reply lib/src/platform_specific/platform_browser_interface.dart]: Well, it's true. But may as well admit it's `navigator.appVersion`.
I'm just open to using something better if we can find it.

File: example/platform_using_switch.dart (hunk 4 above)
Comment (lrhn) [reply example/platform_using_switch.dart:1]: Good idea.

File: lib/platform.dart (hunk 3 above)
Comment (lrhn) [reply lib/platform.dart]: Will do.

File: lib/platform.dart (hunk 2 above)
Comment (lrhn) [reply lib/platform.dart]: ~~I think the "a" makes it parse better. I stumble reading it with. Could be a "the" instead.~~
... unless you read the next line and sees it's plural. Done.

File: lib/platform.dart (hunk 1 above)
Comment (lrhn) [reply lib/platform.dart]: Very. Probably just a surplus "the".
But "can" is not a good word for a DartDoc. Do or do not (and say when, if it's both).

**Comment by mosuem** [comment]
Closing as the [dart-lang/platform](https://github.com/dart-lang/platform) repository is merged into the [dart-lang/core](https://github.com/dart-lang/core) monorepo. Please re-open this PR there!

**Pull request closed without merging**